File: sdk/ruby/test/client_test.rb
require "minitest/autorun"
require "minitest/mock"
require_relative "../lib/microscope_client"

class ClientTest < Minitest::Test
  def setup
    @client = MicroscopeClient::Client.new(base_url: "http://localhost:8093/microscope/")
  end

  def test_record_posts_name_and_content
    response = Minitest::Mock.new
    response.expect(:is_a?, true, [Net::HTTPSuccess])
    response.expect(:body, JSON.generate({ id: "entry-1" }))

    http = Minitest::Mock.new
    http.expect(:use_ssl=, nil, [false])
    http.expect(:read_timeout=, nil, [5])
    http.expect(:request, response, [Net::HTTP::Post])

    Net::HTTP.stub(:new, http) do
      id = @client.record("payment_charged", content: { amount: 4200 })
      assert_equal "entry-1", id
    end

    http.verify
  end

  def test_raises_on_non_success_response
    response = Minitest::Mock.new
    response.expect(:is_a?, false, [Net::HTTPSuccess])
    response.expect(:code, "500")

    http = Minitest::Mock.new
    http.expect(:use_ssl=, nil, [false])
    http.expect(:read_timeout=, nil, [5])
    http.expect(:request, response, [Net::HTTP::Post])

    Net::HTTP.stub(:new, http) do
      assert_raises(MicroscopeClient::Error) do
        @client.record("boom")
      end
    end
  end
end

File: sdk/typescript/test/client.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import { MicroscopeClient } from "../src/index.ts";

interface Call {
  url: string;
  init?: RequestInit;
}

function fakeFetch(response: unknown, ok = true, status = 200) {
  const calls: Call[] = [];
  const fetchImpl = (async (url: string, init?: RequestInit) => {
    calls.push({ url, init });
    return {
      ok,
      status,
      json: async () => response,
    } as Response;
  }) as typeof fetch;
  return { fetchImpl, calls };
}

test("record() posts name and content, returns id", async () => {
  const { fetchImpl, calls } = fakeFetch({ id: "entry-1" });
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope/", fetchImpl });

  const id = await client.record("payment_charged", { amount: 4200 });

  assert.equal(id, "entry-1");
  assert.equal(calls.length, 1);
  assert.equal(calls[0].url, "http://localhost:8093/microscope/api/entries");
  assert.equal(calls[0].init?.method, "POST");
  assert.deepEqual(JSON.parse(calls[0].init?.body as string), {
    name: "payment_charged",
    content: { amount: 4200 },
  });
});

test("record() defaults content to an empty object", async () => {
  const { fetchImpl, calls } = fakeFetch({ id: "entry-2" });
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope", fetchImpl });

  await client.record("no_content_event");

  assert.deepEqual(JSON.parse(calls[0].init?.body as string), {
    name: "no_content_event",
    content: {},
  });
});

test("record() throws on a non-ok response", async () => {
  const { fetchImpl } = fakeFetch({}, false, 500);
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope", fetchImpl });

  await assert.rejects(() => client.record("boom"), /status 500/);
});

test("listEntries() builds query params and skips undefined values", async () => {
  const { fetchImpl, calls } = fakeFetch({ entries: [], total: 0 });
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope", fetchImpl });

  await client.listEntries({ type: "custom", limit: 20 });

  assert.equal(calls[0].url, "http://localhost:8093/microscope/api/entries?type=custom&limit=20");
});

test("getEntry() builds the correct url", async () => {
  const { fetchImpl, calls } = fakeFetch({ id: "entry-1" });
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope", fetchImpl });

  const entry = await client.getEntry("entry-1");

  assert.equal(calls[0].url, "http://localhost:8093/microscope/api/entries/entry-1");
  assert.deepEqual(entry, { id: "entry-1" });
});

test("getEntry() throws on a non-ok response", async () => {
  const { fetchImpl } = fakeFetch({}, false, 404);
  const client = new MicroscopeClient({ baseUrl: "http://localhost:8093/microscope", fetchImpl });

  await assert.rejects(() => client.getEntry("missing"), /status 404/);
});
